### PR TITLE
Add linux-command task schemas and validate task default value using faked options

### PR DESCRIPTION
**Pull request by yyscamper**
Though a lot file change in this PR, but mostly they are similar, all changes are divided into following catalogs:

- Add task schemas for all linux-command related tasks
- Validate task default values using auto-generated faked options. (see changes in spec/lib/task-data/tasks/base-tasks-spec.js, and the library [json-schema-faker](https://github.com/json-schema-faker/json-schema-faker))
- Change task schema unit-test to use filename rather than filepath. (see changes in spec/lib/task-data/schemas/install-os-schema-ut-helper.js)

Related Stories:
- https://www.pivotaltracker.com/n/projects/1492892/stories/126372425
- https://www.pivotaltracker.com/n/projects/1492892/stories/126194769

@RackHD/corecommitters @WangWinson

### CODE REVIEW

File: lib/task-data/schemas/download-file.json [hunk 1]
```diff
@@ -0,0 +1,60 @@
+{
+    "$schema": "rackhd-task-schema.json",
+    "copyright": "Copyright 2016, EMC, Inc.",
+    "title": "Firmware Update",
+    "description": "The schema for firmware update task",
+    "describeJob": "Job.Linux.Commands",
+    "definitions": {
+        "File": {
+            "description": "The filename that to download",
```
Comment (benbp) [inline lib/task-data/schemas/download-file.json]: Can get rid of "that" here.

File: lib/task-data/schemas/download-file.json [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            "type": "string",
+            "minLength": 1
+        },
+        "DownloadDir": {
+            "description": "The microkernel directory where the file will be downloaded to",
+            "type": "string",
+            "minLength": 1
+        },
+        "FileMd5Uri": {
+            "description": "The URL that return the MD5 of the target file",
```
Comment (benbp) [inline lib/task-data/schemas/download-file.json]: "that returns"

File: lib/task-data/schemas/download-file.json [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            "type": "string",
+            "minLength": 1,
+            "pattern": "^https?://"
+        },
+        "FileUri": {
+            "description": "The URL that to download the target file",
```
Comment (benbp) [inline lib/task-data/schemas/download-file.json]: "The URL that is used to download the target file" maybe?

File: lib/task-data/schemas/linux-command.json [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                    "description": "Indicate the command result should be stored into catalog and specify the parameter while storing catalog",
+                    "type": "object",
+                    "properties": {
+                        "source": {
+                            "description": "The source where the catalog data is from",
+                            "type": "string",
+                            "minLength": 1
+                        },
+                        "format": {
+                            "description": "The catalog format, such as 'json', 'raw', .etc",
```
Comment (benbp) [inline lib/task-data/schemas/linux-command.json]: Is it possible to make this an enum?

File: lib/task-data/schemas/linux-command.json [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                        }
+                    },
+                    "required": ["source"],
+                    "additionalProperties": false
+                },
+                "command": {
+                    "$ref": "#/definitions/SimpleCommandItem"
+                },
+                "retries": {
+                    "description": "The totoal retry times if the command execution fails",
```
Comment (benbp) [inline lib/task-data/schemas/linux-command.json]: "total"

**Comment by benbp** [comment]
A couple minor things, but 👍 this is awesome!!

**Comment by WangWinson** [comment]
A new record ? There are 92 file changes !
It gonna take me some time to review .

File: lib/task-data/schemas/linux-command.json (hunk 4 above)
Comment (yyscamper) [inline lib/task-data/schemas/linux-command.json]: good idea.

Comment (yyscamper) [inline lib/task-data/schemas/linux-command.json]: Done

**Comment by cgx027** [comment]
Fabulous! Looks good to me :+1:

File: lib/task-data/schemas/download-file.json [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            "description": "The URL that returns the MD5 of the target file",
+            "type": "string",
+            "minLength": 1,
+            "pattern": "^https?://"
+        },
+        "FileUri": {
+            "description": "The URL that is used to download the target file",
+            "type": "string",
+            "minLength": 1,
+            "pattern": "^https?://"
```
Comment (WangWinson) [inline lib/task-data/schemas/download-file.json]:
```JSON
"format": "uri"
```
is not enough here ?

File: lib/task-data/schemas/linux-command.json [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                },
+                "retries": {
+                    "description": "The total retry times if the command execution fails",
+                    "type": "integer",
+                    "minimum": 0
+                },
+                "downloadUrl": {
+                    "description": "Specify the download URL (relative to RackHD southbound API server) if the command needs to download some data or script from RackHD before execution",
+                    "type": "string",
+                    "pattern": "^/[a-zA-Z0-9+-.#?=\\\/]+"
```
Comment (WangWinson) [inline lib/task-data/schemas/linux-command.json]: so, this url will not start with `http://` ?

File: spec/lib/task-data/schemas/set-bmc-credentials-spec.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        password: 'myPassword'
+    };
+
+    var positiveSetParam = {
+        user: ['a', '1122334455667788'],
+        password: ['a', '11223344556677889900']
+    };
+
+    var negativeSetParam = {
+        user: ['', null, '11223344556677889', 'a测试b', '~そうですね', '♥'],
```
Comment (WangWinson) [inline spec/lib/task-data/schemas/set-bmc-credentials-spec.js]: wow, where is google translate

**Comment by iceiilin** [comment]
Great work. :+1:

**Comment by WangWinson** [comment]
:+1:  Looks awesome in general

File: lib/task-data/schemas/linux-command.json (hunk 7 above)
Comment (yyscamper) [inline lib/task-data/schemas/linux-command.json]: As I said in  the description, this is a URL  that relative to RackHD southboud API server, below is an example:
https://github.com/RackHD/on-tasks/blob/ed2a5e49998a04da0901319f015faec841c19bba/lib/task-data/tasks/catalog-smart.js#L13

File: lib/task-data/schemas/download-file.json (hunk 6 above)
Comment (yyscamper) [inline lib/task-data/schemas/download-file.json]: Actually, I tested the `abc/` is still a valid uri format, but this is not what we want. For a format uri is not enough here.

File: spec/lib/task-data/schemas/set-bmc-credentials-spec.js (hunk 8 above)
Comment (benbp) [inline spec/lib/task-data/schemas/set-bmc-credentials-spec.js]: Amazing

**Pull request merged**